# Joann Chan

______

## Project B

**We will be taking a look at the
[*62 Amazing Facts You Probably Didn't Know That Will Blow Your Mind*](https://www.buzzfeed.com/readcommentbackwards/62-amazing-fact-you-probably-didnt-know-that-will-dmjk?utm_term=.cfpEWdn4R#.ktYgO5XAL) article from Buzzfeed.**

_______

In [1]:
import requests
top = requests.get('https://www.buzzfeed.com/readcommentbackwards/62-amazing-fact-you-probably-didnt-know-that-will-dmjk?utm_term=.cfpEWdn4R#.ktYgO5XAL')

In [2]:
print(top.text[:100])

<!DOCTYPE html>
<html lang="en" class="core-ux" data-app='bpager' data-env=prod>
    <head>
        


In [3]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(top.text, 'html.parser')

**Let's start by tracking down the data we want from the website, the 62 facts**

**Let's take a look at the first interesting fact:**
> ***The bushes and clouds in Super Mario Bros are the same, just colored differently.***

**You want to copy this fact for what we are about to do next**

**Right-click on the webpage and scroll to the bottom to select:**
>***View Page Source*** 

**This should take you to the article's view-source page, which holds all the codes that make up that webpage**

**Now, on the view-source page, press *Ctrl + F***

**The *Find* toolbar should pop up**

**Now, remember when we copied that interesting fact?**

**Paste that fact into the *Find* toolbar**

**What we should get is the line of code that writes up that first interesting fact:**
> $<span class="js-subbuzz__title-text">The bushes and clouds in Super Mario Bros are the same, just colored differently.</span>$
>>**Also, take note of which line this code is written on because we will need to use this information later on.**

>>>**We will need to look at the left side of the screen where we see numbers going down in order on a vertical grey bar.**

>>>**We see that the code to writeup the first interesting fact is on line 1638**

**Now back to the code itself**

**Notice that the codes surrounding the text are:**
> ***$<span class="js-subbuzz__title-text">$  and  $</span>$***
>> **When we copy and paste other facts into the *Find* toolbar to find them in the view-source page, we will see that they all are surrounded with the same code, therefore we can conclude that all 62 interesting facts will be surrounded by $<span class="js-subbuzz__title-text">$  and  $</span>$**

**This allows us to write up the following code:**
> ***facts = soup.find_all('span')***
>> **This is essentially saying to set the word *"facts"* equal to finding all the texts that are surrounded by the code $<span>$**

____

**I REALIZED THAT THE CODE I USED:**

> ***facts = soup.find_all('span')***

**GAVE ME TOO LARGE OF A RANGE OF CODES FOR ONLY 62 FACTS SO I WENT BACK INTO VIEW PAGE SOURCE AND REALIZED A BETTER SOLUTION TO FINDING THE FACTS, BUT I AM JUST GOING TO LEAVE EVERYTHING I DID WITH THIS CODE BELOW SO THAT I CAN REFER TO IT IN THE FUTURE IF I NEED IT. SORRY!**

**I AM GOING TO PUT *h3* INTO *facts = soup.find_all('')* BECAUSE THIS WILL ACTUALLY GIVE ME ALL THE FACTS WITHOUT ANY EXTRA INFORMATION**


**IF YOU WANT TO SKIP PAST TO WHERE I USE *h3* INSTEAD OF *span***
# [PRESS HERE](#using-h3)

_____

In [4]:
facts = soup.find_all('span')

In [5]:
len(facts)

1131

**We can see that there are 1131 facts, but we know that there should only be 62.**

**The reason why this number is larger than it should be is due to the fact that there are other texts, not just the interesting facts, that are surrounded by the same code of $<span>$**

**If we type in $</span>$ into the *Find* toolbar, we can find all the texts that end with that code, which should allow us to see where Interesting Fact 1 starts**
> **Remember when I told you to find which line the first interesting fact was written on? Line 1638?**

>**Well, now we need to find that line again because we typed in a new search into the *Find* toolbar, which takes us away from were we need to be**

>**What we do is scroll through the page, looking at that vertical grey bar, and find the number 1638, which should bring us back to the first interesting fact**

>>**You might be wondering why we need to go through all these steps when we were already at the first fact.**

>>**Well, we need to do this because we want to find which number, out of the 1131 codes, is the first of the interesting facts.**

>>**The great thing with the *Find* toolbar is that it tells us which $</Span>$ we are at out of all 1131 options.**

>>**The only problem is that we may have lost our place and we can't navigate our way back to the $</Span>$ that has our first interesting fact, which is why we needed to know which line of code it was written on.**

>>**The only other way to do this would be to guess and check, which is too tedious.**

>**We can see that, out of the 1131 codes, the first interesting fact is actually the 37th code**

>**Let's check to be sure by writing a code**



In [6]:
facts[37]

<span class="icon icon--primary">
<svg>
<title>Share On facebook</title>
      Share On facebook 
      <use xlink:href="#facebook"></use>
</svg>
</span>

**This doesn't seem to be the correct text, so lets try a number close to 37, like 36**

In [7]:
facts[36]

<span class="js-subbuzz__title-text">The bushes and clouds in Super Mario Bros are the same, just colored differently.</span>

**Is this the correct code we were looking for? YES!**

**Now let's try to remove the code surrounding the text so that we would only see the fact**

In [8]:
facts[36].text

'The bushes and clouds in Super Mario Bros are the same, just colored differently.'

**Did we get the text by itself? We sure did!**

**Ok, so we found the first fact, but we want to find all 62. So, let's go back to the view source page and locate the last code using the same method we used to find fact 1**

**We know that the fact is:**
>**A banana is actually a berry. A strawberry isn't.**

**And the code we found was:**
>$ <span class="js-subbuzz__title-text">A banana is actually a berry. A strawberry isn&#39;t.</span>$

**Which is on line 21326**

**Which is #1074 of the 1131 codes**

In [9]:
facts[1073].text

"A banana is actually a berry. A strawberry isn't."

**Again, we were 1 number off from the actual code**

**But, now we know that the facts lay somewhere on and between 36 to 1073**

In [10]:
facts = facts[36:1073]

_____

# Using h3 <a id="using-h3"></a>

_________

In [11]:
facts = soup.find_all('h3')

In [12]:
len(facts)

63

**We can see that this is a much closer to 62 than 1131**

**We are essentially going to do everything the same as before, but omitting any unnecessary steps that we may have taken before**

In [13]:
facts[0]

<h3 class="subbuzz__title xs-mb1 bold">
<span class="subbuzz__number">1.</span>
<span class="js-subbuzz__title-text">The bushes and clouds in Super Mario Bros are the same, just colored differently.</span>
</h3>

In [14]:
facts[0].text

'\n1.\nThe bushes and clouds in Super Mario Bros are the same, just colored differently.\n'

**We see that 0 is were we get out first interesting fact**

**Lets try to see where we can get our last fact**

In [15]:
facts[61].text

"\n62.\nA banana is actually a berry. A strawberry isn't.\n"

**YES! This is our last fact, so we know that the facts are from 0 to 61.**

In [16]:
facts = facts[0:62]

**We put 62 instead of 61 because this code will omit the last number**

# Creating a Table for the Facts

____

In [17]:
table = []
for i in facts:
    fact = i.text
    table.append(fact)

In [18]:
len(table)

62

In [19]:
table[:62]

['\n1.\nThe bushes and clouds in Super Mario Bros are the same, just colored differently.\n',
 '\n2.\nThere are warehouses so big they have their own WEATHER.\n',
 '\n3.\nIn 1903 the Wright brothers flew for 59 seconds. 38 years later the Japanese bombed Pearl Harbor. 28 years after that, we landed on the moon.\n',
 '\n4.\nWhen a male bee climaxes, their testicles explode then they die.\n',
 '\n5.\nCleopatra lived closer in time to the first Moon landing than to the building of the Great Pyramid.\n',
 '\n6.\nDuck Hunt is two player. A controller in port 2 controls the ducks.\n',
 '\n7.\nThere are more fake flamingos in the world than real flamingos.\n',
 '\n8.\nThe adult human has two to nine pounds of bacteria in his or her body.\n',
 '\n9.\nBlue whales heart is the size of a VW Beetle and that you could swim through some of its arteries.\n',
 '\n10.\nHydrogen is a light, odorless gas, which, given enough time, turns into people.\n',
 '\n11.\nAll of the gold mined in the history of th

**Here we can see all of the 62 facts**

In [20]:
import pandas as pd
df = pd.DataFrame(table, columns=['Interesting Facts'])

In [21]:
df

Interesting Facts
0   \n1.\nThe bushes and clouds in Super Mario Bro...
1   \n2.\nThere are warehouses so big they have th...
2   \n3.\nIn 1903 the Wright brothers flew for 59 ...
3   \n4.\nWhen a male bee climaxes, their testicle...
4   \n5.\nCleopatra lived closer in time to the fi...
5   \n6.\nDuck Hunt is two player. A controller in...
6   \n7.\nThere are more fake flamingos in the wor...
7   \n8.\nThe adult human has two to nine pounds o...
8   \n9.\nBlue whales heart is the size of a VW Be...
9   \n10.\nHydrogen is a light, odorless gas, whic...
10  \n11.\nAll of the gold mined in the history of...
11  \n12.\nThere are more atoms in a single glass ...
12  \n13.\nHalf of all humans who have ever lived,...
13  \n14.\nMartin Sheen's original name was Ramon ...
14  \n15.\nAlcohol is considered a performance-enh...
15  \n16.\nOnce Charlie Chaplin entered a contest ...
16  \n17.\nIf you shuffle a deck of cards, chances...
17  \n18.\nOf all the people in history that have ...
18  \n19.\nFold over a piece of paper 42 times and...
19  \n20.\nPirates wear eye patches so that they c...
20  \n21.\nIt took thousands of years after domest...
21  \n22.\n54 million people alive right now will ...
22  \n23.\nYou replace every particle in your body...
23  \n24.\nWhen your mother was born, she was alre...
24  \n25.\nA tree is the opposite of your lungs. P...
25     \n26.\nMayonnaise is made from oil and eggs.\n
26  \n27.\nThe sun is 400x further from the earth ...
27  \n28.\nAll the humans on the planet could fit ...
28            \n29.\nGoats have rectangular pupils.\n
29  \n30.\nIf there was no space between any of it...
..                                                ...
32  \n33.\nSloths mistakenly grab their arms inste...
33  \n34.\nPenguins will give their mate a pebble ...
34                \n35.\nButterflies are cannibals.\n
35    \n36.\nTurtles can breathe out of their butts\n
36  \n37.\nA small percentage of static on televis...
37  \n38.\nIt takes a photon, on average, 200,000 ...
38  \n39.\nBoth of Jack Black's parents were rocke...
39           \n40.\nA pigs orgasm lasts 30 minutes.\n
40  \n41.\nVelociraptors were just slightly bigger...
41  \n42.\nA pencil has the potential to draw a li...
42  \n43.\nThere is a species of jellyfish that is...
43  \n44.\nThe largest air force in the world is t...
44  \n45.\nHorseshoe crabs have eyes on their tail.\n
45  \n46.\nA Nintendo had over twice the computing...
46    \n47.\nLighters were invented before matches.\n
47  \n48.\nPsychology is the brain trying to compr...
48  \n49.\nIf a man never cut his beard, by the ti...
49  \n50.\n1 sperm has 37.5MB of DNA information i...
50  \n51.\nThe population of Ireland is still 2 mi...
51  \n52.\nThe last president that wasn't a Republ...
52  \n53.\nYou have 0.5% chance of being related t...
53  \n54.\nThe USA is an older country than German...
54  \n55.\nNapoleon was not unusually short. The F...
55  \n56.\nIf you keep going North, you will event...
56  \n57.\nVending machines kill 4 times as many p...
57  \n58.\nThere is a garbage swirl in the ocean t...
58  \n59.\nBlack HP printer ink costs roughly 1,40...
59  \n60.\nGeorge Washington spent about 7% of his...
60                  \n61.\nA Mosquito has 47 teeth.\n
61  \n62.\nA banana is actually a berry. A strawbe...

[62 rows x 1 columns]

**Now that we have all our facts in a table, it is much more pleasant to look at and much easier for us to locate each fact**